### PR TITLE
Build navigation into JSON

### PR DESCRIPTION
**Pull request by sknorr**
The current state is a bit WIP, because it does not actually copy the HTML/PHP template, it just generates JSON files. the part with copying everything is easy though (even if you do it manually).

### CODE REVIEW

**Comment by sknorr** [comment]
@tomschr i will merge this immediately because it helps with our current build testing a lot, but if you have time, could you look through it please?

File: bin/docserv-build-navigation [hunk 1]
```diff
@@ -119,23 +122,24 @@ for file in $template_html $template_translation $template_resources; do
...
+# Where to place the JSON data files
+data_path='docserv/data'
+# Where to place the template's resource files (JS, CSS, images)
+res_path='docserv/res'
+for product in $allproducts; do
+  mkdir -p $output_dir/$data_path/$product
+done
+for lang in "$ui_languages"; do
+  for product in $allproducts; do
+    mkdir -p $output_dir/$lang/$product
```
Comment (tomschr) [inline bin/docserv-build-navigation]: The two for loops in line 132 and 135 could be merged like this:

```bash
for lang in "$ui_languages"; do
  for product in $allproducts; do
    mkdir -p $output_dir/$lang/$product
    mkdir -p $output_dir/$data_path/$product
done
```

File: share/build-navigation/cache-request.xsl [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    <xsl:param name="setid" select="''"/>
+    <xsl:param name="doc_language" select="''"/>
+    <xsl:param name="dc" select="''"/>
+    <xsl:param name="rootid" select="''"/>
+
+    <xsl:variable name="title">
+      <xsl:choose>
+        <xsl:when
+          test="$rootid != '' and
+                ($cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]/title[@rootid = $rootid])[1]">
```
Comment (tomschr) [inline share/build-navigation/cache-request.xsl]: These long variables seems a bit often repeated. Could you extract the part that is "constant" to all? Something like this:

```xml
<!-- maybe use a better name for the variable -->
<xsl:variable name="content" select="$cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]"/

<xsl:variable name="title">
   <xsl:choose>
      <xsl:when="$rootid != '' and ($content/title[@rootid = $rootid])[1]">
         <!-- ... -->
      </xsl:when>
   </xsl:choose>
</xsl:variable>
```

I think, that is better readable and should speed up things.

File: bin/docserv-build-navigation (hunk 1 above)
Comment (tomschr) [reply bin/docserv-build-navigation]: Oh wait, maybe this is better:

```bash
for product in $allproducts; do
   mkdir -p $output_dir/$data_path/$product
   for lang in "$ui_languages"; do
      mkdir -p $output_dir/$lang/$product
   done
done
```

I haven't tried it, it's just a small optimization idea. Use it or leave it. :wink:

Comment (sknorr) [reply bin/docserv-build-navigation]: Not quite, for `$data_path`, the `$ui_language` is irrelevant.

Comment (tomschr) [reply bin/docserv-build-navigation]: Yes, I know, but it doesn't matter. :-) 
You have two for loops which iterate two times over `$allproducts`. My suggestion takes these two separate for loops together so you don't need to repeat yourself. Saves you some CPU cycles. ;-)

**Pull request merged**